metrics/src/metrics.rs

```diff
@@ -288,16 +291,40 @@ impl MetricsAgent {
         trace!("run: exit");
     }
 
+    fn send(&self, command: MetricsCommand) {
+        let try_send = |command| {
+            match self.sender.try_send(command) {
+                Err(TrySendError::Full(cmd)) => {
+                    warn!("metric queue full, point dropped: {:?}.", cmd);
+                }
+                Err(TrySendError::Disconnected(_)) => {
+                    warn!("metric queue disconnected.");
+                }
+                // success
+                Ok(_) => {}
+            }
+        };
+
+        // Report a datapoint for the metric command queue size when the queue
+        // is filling up
+        let len = self.sender.len();
+        if len > METRIC_COMMANDS_QUEUE_LENGTH_REPORT_THRESHOLD {
+            try_send(MetricsCommand::Submit(
+                DataPoint::new("metric_command_queue")
+                    .add_field_i64("len", len as i64)
+                    .to_owned(),
+                Level::Info,
+            ));
+        }
```

Comment:
After channel gets bigger than the threshold, with each data-point this is sending an additional data-point down the channel. i.e. right at the time that channel is already overloaded, this is doubling the number of metrics sent down the channel. 

Also we don't need to track the size of the channel each time a data-point is submitted.
It suffices to inspect the size of the channel only when there is a write here:
https://github.com/solana-labs/solana/blob/099258e4e/metrics/src/metrics.rs#L277-L286

Reply:
Make sense. Updated.